pkg/agent/controller/packetsampling/packetsampling_controller.go

```diff
@@ -0,0 +1,651 @@
+		if len(dstPodInterfaces) > 0 {
+			if packet.IsIPv6 {
+				packet.DestinationIP = dstPodInterfaces[0].GetIPv6Addr()
+			} else {
+				packet.DestinationIP = dstPodInterfaces[0].GetIPv4Addr()
+			}
+		} else {
+			dstPod, err := c.kubeClient.CoreV1().Pods(ps.Spec.Destination.Namespace).Get(context.TODO(), ps.Spec.Destination.Pod, metav1.GetOptions{})
+			if err != nil {
+				return nil, fmt.Errorf("failed to get the destination pod %s/%s: %v", ps.Spec.Destination.Namespace, ps.Spec.Destination.Pod, err)
+			}
+			podIPs := make([]net.IP, len(dstPod.Status.PodIPs))
+			for i, ip := range dstPod.Status.PodIPs {
+				podIPs[i] = net.ParseIP(ip.IP)
+			}
+			if packet.IsIPv6 {
+				packet.DestinationIP, _ = util.GetIPWithFamily(podIPs, util.FamilyIPv6)
+			} else {
+				packet.DestinationIP = util.GetIPv4Addr(podIPs)
+			}
+		}
+		if packet.DestinationIP == nil {
+			if packet.IsIPv6 {
+				return nil, errors.New("destination Pod does not have an IPv6 address")
+			}
+			return nil, errors.New("destination Pod does not have an IPv4 address")
+		}
+	} else if ps.Spec.Destination.Service != "" {
+		dstSvc, err := c.serviceLister.Services(ps.Spec.Destination.Namespace).Get(ps.Spec.Destination.Service)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get the destination service %s/%s: %v", ps.Spec.Destination.Namespace, ps.Spec.Destination.Service, err)
+		}
+		if dstSvc.Spec.ClusterIP == "" {
+			return nil, errors.New("destination Service does not have an ClusterIP")
+		}
+
+		packet.DestinationIP = net.ParseIP(dstSvc.Spec.ClusterIP)
+		if !packet.IsIPv6 {
+			packet.DestinationIP = packet.DestinationIP.To4()
+			if packet.DestinationIP == nil {
+				return nil, errors.New("destination Service does not have an IPv4 address")
+			}
+		} else if packet.DestinationIP.To4() != nil {
+			return nil, errors.New("destination Service does not have an IPv6 address")
+		}
+	} else {
+		return nil, errors.New("destination is not specified")
+	}
+
+	if ps.Spec.Packet.TransportHeader.TCP != nil {
+		packet.SourcePort = uint16(ps.Spec.Packet.TransportHeader.TCP.SrcPort)
+		packet.DestinationPort = uint16(ps.Spec.Packet.TransportHeader.TCP.DstPort)
+		if ps.Spec.Packet.TransportHeader.TCP.Flags != 0 {
+			packet.TCPFlags = uint8(ps.Spec.Packet.TransportHeader.TCP.Flags)
+		}
+	} else if ps.Spec.Packet.TransportHeader.UDP != nil {
+		packet.SourcePort = uint16(ps.Spec.Packet.TransportHeader.UDP.SrcPort)
+		packet.DestinationPort = uint16(ps.Spec.Packet.TransportHeader.UDP.DstPort)
+	}
+
+	packet.IPProto = parseTargetProto(&ps.Spec.Packet)
+	return packet, nil
+}
+
+func parseTargetProto(packet *crdv1alpha1.Packet) uint8 {
+	var ipProto uint8
+	var isIPv6 bool
+	if packet.IPv6Header != nil {
+		isIPv6 = true
+		if packet.IPv6Header.NextHeader != nil {
+			ipProto = uint8(*packet.IPv6Header.NextHeader)
+		}
+	} else if packet.IPHeader.Protocol != 0 {
+		ipProto = uint8(packet.IPHeader.Protocol)
+	}
+
+	if packet.TransportHeader.TCP != nil {
```

Comment:
I suppose you need to validate the `packet.TransportHeader.*` and `packet.IPv6Header.NextHeader` to see they are matched or not?

Reply:
OK. will add the check to the validation webhook.

Reply:
updated